internal/glfw/gen.go

```diff
@@ -0,0 +1,171 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build ignore
+// +build ignore
+
+package main
+
+import (
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+)
+
+// files that should be compiled
+var filenames = map[string]struct{}{
+	"cocoa_init.m":     {},
+	"cocoa_joystick.m": {},
+	"cocoa_monitor.m":  {},
+	"cocoa_time.c":     {},
+	"cocoa_window.m":   {},
+	"context.c":        {},
+	"egl_context.c":    {},
+	"init.c":           {},
+	"input.c":          {},
+	"monitor.c":        {},
+	"nsgl_context.m":   {},
+	"osmesa_context.c": {},
+	"posix_thread.c":   {},
+	"vulkan.c":         {},
+	"window.c":         {},
+}
+
+type arch string
+
+const (
+	archAmd64 arch = "amd64"
+	archArm64 arch = "arm64"
+)
+
+func (a arch) target() string {
+	switch a {
+	case archAmd64:
+		return "x86_64"
+	case archArm64:
+		return "arm64"
+	default:
+		panic("not reached")
+	}
+}
+
+func execCommand(name string, args ...string) error {
+	cmd := exec.Command(name, args...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	return cmd.Run()
+}
+
+func objectExt(s string) string {
+	return strings.Trim(strings.Trim(s, ".c"), ".m") + ".o"
+}
+
+func run() error {
+	defer log.Printf("Finished!\n")
+	list := exec.Command("go", "list", "-f", "{{.Dir}}", "github.com/go-gl/glfw/v3.3/glfw")
+	output, err := list.Output()
+	if err != nil {
+		return err
+	}
+	source := strings.TrimSpace(string(output))
+	build, err := os.MkdirTemp("", "ebitengine-glfw-*")
+	if err != nil {
+		return err
+	}
+	defer func() {
+		log.Printf("Deleting build files: %s", build)
+		os.RemoveAll(build) // clean up
+	}()
+	csource := source + "/glfw/src"
+	includes := source + "/glfw/include"
+	dir, err := os.ReadDir(csource)
+	if err != nil {
+		return err
+	}
+
+	for _, a := range []arch{archAmd64, archArm64} {
+		log.Printf("Compiling Files for %s at %s\n", a.target(), build)
+		for _, entry := range dir {
+			if _, ok := filenames[entry.Name()]; !ok {
+				continue
+			}
+			name := entry.Name()
+			args := []string{
+				"-o", // output
+				objectExt(filepath.Join(build, name)),
+				"-mmacosx-version-min=10.12",
+				"-arch",
+				a.target(),
+				"-c", // compile without linking
+				filepath.Join("-I", includes, "include/GLFW/glfw3.h"),       // include glfw3.h
+				filepath.Join("-I", includes, "include/GLFW/glfw3native.h"), // include glfw3native.h
+				"-D_GLFW_COCOA",              // define _GLFW_COCOA
+				"-D_GLFW_BUILD_DLL",          // define _GLFW_BUILD_DLL to build dylib
+				"-Os",                        // optimize for size
+				"-g",                         // debug info
+				filepath.Join(csource, name), // .c or .m file
```

Comment:
Is this building now deterministic?

Reply:
I'm not sure what is required for it to be deterministic.

Reply:
I was wondering if you have confirmed that the build result was always the same

Reply:
I followed http://blog.llvm.org/2019/11/deterministic-builds-with-clang-and-lld.html as a guide.

Reply:
Like exactly the same binary result each time?

Reply:
> Like exactly the same binary result each time?

Yes

Reply:
This property is necessary since unnecessary commit changes would be generated whenever `go generate` was executed.

Reply:
I see. I am not getting deterministic results. I'll try to figure out why

Reply:
Me netither. The result doesn't seem deterministic yet.